### ISSUE TITLE
Errors installing dlesym

### ISSUE THREAD

**Issue by matthewjaniga** [NVIDIA/earth2studio#423]
Greetings,

When attempting to add dlesym I am getting the following error.

uv add earth2studio --extra dlesym

I get similar issues when installing via the pip method.

Resolved 128 packages in 3ms
  × Failed to build `earth2grid @ git+https://github.com/NVlabs/earth2grid.git@aefb10793aa372bae7d0951d627a89e2983fd0ca`
  ├─▶ The build backend returned an error
  ╰─▶ Call to `setuptools.build_meta.build_wheel` failed (exit status: 1)

      [stdout]
      Detected GNU compiler.
      running bdist_wheel
      running build
      running build_py
      running egg_info
      writing earth2grid.egg-info/PKG-INFO
      writing dependency_links to earth2grid.egg-info/dependency_links.txt
      writing requirements to earth2grid.egg-info/requires.txt
      writing top-level names to earth2grid.egg-info/top_level.txt
      reading manifest file 'earth2grid.egg-info/SOURCES.txt'
      adding license file 'LICENSE.txt'
      writing manifest file 'earth2grid.egg-info/SOURCES.txt'
      running build_ext
      building 'earth2grid._healpix_bare' extension
      CC -fno-strict-overflow -Wsign-compare -DNDEBUG -O2 -Wall -fPIC -O2 -isystem /p/home/janiga/miniforge3/include -fPIC -O2 -isystem /p/home/janiga/miniforge3/include -fPIC
      -I/p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/csrc -I/p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare
      -I/p/home/janiga/.cache/uv/builds-v0/.tmp6flWvL/lib/python3.12/site-packages/torch/include -I/p/home/janiga/.cache/uv/builds-v0/.tmp6flWvL/lib/python3.12/site-packages/torch/include/torch/csrc/api/include
      -I/p/home/janiga/.cache/uv/builds-v0/.tmp6flWvL/include -I/p/home/janiga/miniforge3/include/python3.12 -c earth2grid/csrc/healpix_bare_wrapper.cpp -o
      build/temp.linux-x86_64-cpython-312/earth2grid/csrc/healpix_bare_wrapper.o -std=c++20 -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\"
      -DPYBIND11_BUILD_ABI=\"_cxxabi1016\" -DTORCH_EXTENSION_NAME=_healpix_bare -D_GLIBCXX_USE_CXX11_ABI=1

      [stderr]
      No ROCm runtime is found, using ROCM_HOME='/opt/rocm'
      <string>:84: UserWarning: Cuda extensions for torch not found, skipping cuda healpix padding module
      /p/home/janiga/.cache/uv/builds-v0/.tmp6flWvL/lib/python3.12/site-packages/setuptools/config/_apply_pyprojecttoml.py:82: SetuptoolsDeprecationWarning: `project.license` as a TOML table is deprecated
      !!

              ********************************************************************************
              Please use a simple string containing a SPDX expression for `project.license`. You can also use `project.license-files`. (Both options available on setuptools>=77.0.0).

              By 2026-Feb-18, you need to update your project and remove deprecated calls
              or your builds will no longer be supported.

              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************

      !!
        corresp(dist, value, root_dir)
      /p/home/janiga/.cache/uv/builds-v0/.tmp6flWvL/lib/python3.12/site-packages/setuptools/config/_apply_pyprojecttoml.py:61: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!

              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:

              License :: OSI Approved :: BSD License

              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************

      !!
        dist._finalize_license_expression()
      /p/home/janiga/.cache/uv/builds-v0/.tmp6flWvL/lib/python3.12/site-packages/setuptools/dist.py:759: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!

              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:

              License :: OSI Approved :: BSD License

              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************

      !!
        self._finalize_license_expression()
      /p/home/janiga/.cache/uv/builds-v0/.tmp6flWvL/lib/python3.12/site-packages/torch/utils/cpp_extension.py:576: UserWarning: Attempted to use ninja as the BuildExtension backend but we could not find
      ninja.. Falling back to using the slow distutils backend.
        warnings.warn(msg.format('we could not find ninja.'))
      In file included from earth2grid/csrc/healpix_bare_wrapper.cpp:20:
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:68:47: error: non-constant-expression cannot be narrowed from type 'int64_t' (aka
      'long') to 'int32_t' (aka 'int') in initializer list [-Wc++11-narrowing]
         68 |   return (loc.z >= 0) ? (t_hpc){1.-jm, 1.-jp, ntt}
            |                                               ^~~
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:68:47: note: insert an explicit cast to silence this issue
         68 |   return (loc.z >= 0) ? (t_hpc){1.-jm, 1.-jp, ntt}
            |                                               ^~~
            |                                               static_cast<int32_t>( )
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:69:41: error: non-constant-expression cannot be narrowed from type 'int64_t' (aka
      'long') to 'int32_t' (aka 'int') in initializer list [-Wc++11-narrowing]
         69 |                       : (t_hpc){jp, jm, ntt+8};
            |                                         ^~~~~
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:69:41: note: insert an explicit cast to silence this issue
         69 |                       : (t_hpc){jp, jm, ntt+8};
            |                                         ^~~~~
            |                                         static_cast<int32_t>( )
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:176:59: error: non-constant-expression cannot be narrowed from type 'int64_t' (aka
      'long') to 'int32_t' (aka 'int') in initializer list [-Wc++11-narrowing]
        176 |   return (t_hpd){compress_bits(p2), compress_bits(p2>>1), pix/npface_};
            |                                                           ^~~~~~~~~~~
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:176:59: note: insert an explicit cast to silence this issue
        176 |   return (t_hpd){compress_bits(p2), compress_bits(p2>>1), pix/npface_};
            |                                                           ^~~~~~~~~~~
            |                                                           static_cast<int32_t>( )
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:218:52: error: non-constant-expression cannot be narrowed from type 'int64_t' (aka
      'long') to 'int32_t' (aka 'int') in initializer list [-Wc++11-narrowing]
        218 |     return (t_hpd) {(ipt-irt)>>1, (-(ipt+irt))>>1, face};
            |                                                    ^~~~
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:218:52: note: insert an explicit cast to silence this issue
        218 |     return (t_hpd) {(ipt-irt)>>1, (-(ipt+irt))>>1, face};
            |                                                    ^~~~
            |                                                    static_cast<int32_t>( )
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:234:52: error: non-constant-expression cannot be narrowed from type 'int64_t' (aka
      'long') to 'int32_t' (aka 'int') in initializer list [-Wc++11-narrowing]
        234 |     return (t_hpd) {(ipt-irt)>>1, (-(ipt+irt))>>1, face};
            |                                                    ^~~~
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:234:52: note: insert an explicit cast to silence this issue
        234 |     return (t_hpd) {(ipt-irt)>>1, (-(ipt+irt))>>1, face};
            |                                                    ^~~~
            |                                                    static_cast<int32_t>( )
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:245:52: error: non-constant-expression cannot be narrowed from type 'int64_t' (aka
      'long') to 'int32_t' (aka 'int') in initializer list [-Wc++11-narrowing]
        245 |     return (t_hpd) {(ipt-irt)>>1, (-(ipt+irt))>>1, face};
            |                                                    ^~~~
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:245:52: note: insert an explicit cast to silence this issue
        245 |     return (t_hpd) {(ipt-irt)>>1, (-(ipt+irt))>>1, face};
            |                                                    ^~~~
            |                                                    static_cast<int32_t>( )
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:266:18: error: type 'double' cannot be narrowed to 'int64_t' (aka 'long') in
      initializer list [-Wc++11-narrowing]
        266 |   return (t_hpd){(tmp.x*nside_), (tmp.y*nside_), tmp.f};
            |                  ^~~~~~~~~~~~~~
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:266:18: note: insert an explicit cast to silence this issue
        266 |   return (t_hpd){(tmp.x*nside_), (tmp.y*nside_), tmp.f};
            |                  ^~~~~~~~~~~~~~
            |                  static_cast<int64_t>( )
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:266:34: error: type 'double' cannot be narrowed to 'int64_t' (aka 'long') in
      initializer list [-Wc++11-narrowing]
        266 |   return (t_hpd){(tmp.x*nside_), (tmp.y*nside_), tmp.f};
            |                                  ^~~~~~~~~~~~~~
      /p/home/janiga/.cache/uv/git-v0/checkouts/1dd0445c5bddaec9/aefb107/earth2grid/third_party/healpix_bare/healpix_bare.c:266:34: note: insert an explicit cast to silence this issue
        266 |   return (t_hpd){(tmp.x*nside_), (tmp.y*nside_), tmp.f};
            |                                  ^~~~~~~~~~~~~~
            |                                  static_cast<int64_t>( )
      earth2grid/csrc/healpix_bare_wrapper.cpp:86:24: error: non-constant-expression cannot be narrowed from type 'int64_t' (aka 'long') to 'int32_t' (aka 'int') in initializer list [-Wc++11-narrowing]
         86 |       t_hpd hpd {x, y, f};
            |                        ^
      earth2grid/csrc/healpix_bare_wrapper.cpp:86:24: note: insert an explicit cast to silence this issue
         86 |       t_hpd hpd {x, y, f};
            |                        ^
            |                        static_cast<int32_t>( )
      earth2grid/csrc/healpix_bare_wrapper.cpp:121:48: error: non-constant-expression cannot be narrowed from type 'long' to 'int32_t' (aka 'int') in initializer list [-Wc++11-narrowing]
        121 |       t_hpc hpc {accessor_x[i], accessor_y[i], accessor_f[i]};
            |                                                ^~~~~~~~~~~~~
      earth2grid/csrc/healpix_bare_wrapper.cpp:121:48: note: insert an explicit cast to silence this issue
        121 |       t_hpc hpc {accessor_x[i], accessor_y[i], accessor_f[i]};
            |                                                ^~~~~~~~~~~~~
            |                                                static_cast<int32_t>( )
      10 errors generated.
      error: command '/opt/cray/pe/craype/2.7.34/bin/CC' failed with exit code 1

      hint: This usually indicates a problem with the package or the build environment.
  help: If you want to add the package regardless of the failed resolution, provide the `--frozen` flag to skip locking and syncing.

**Comment by NickGeneva** [comment]
Hi @matthewjaniga 

To confirm, other models run okay and can run on the GPU?
This line here seems concerning:

```
  <string>:84: UserWarning: Cuda extensions for torch not found, skipping cuda healpix padding module
```

Can you confirm that torch cuda is installed and working for PyTorch?

`uv run python`

```python
import torch
torch.cuda.is_available()
```

**Comment by matthewjaniga** [comment]
I don't have cuda on the current system I'm using, I was attempting to run on CPU only as the DLESyM example had this line...

https://nvidia.github.io/earth2studio/examples/14_dlesym_example.html#sphx-glr-examples-14-dlesym-example-py

device = "cpu" if not torch.cuda.is_available() else "cuda"

It seems to be getting stuck on earth2grid. Is installing and running DLESyM possible on a CPU only system? I can try on a system with GPUs if that's not an option.

**Comment by NickGeneva** [comment]
Looping in @pzharrington for additional input.

Earth2Grid does require a GPU to compile and is needed for the LatLon variant of DLESym since it handles the conversion from lat/lon to hpx grids.

I will mention the arch itself does have CPU supported fallbacks for some HPX operations but it assumes your data is in HPX already. We'll update the example to reflect the GPU requirement. Apologies for the confusion.

**Comment by matthewjaniga** [comment]
I was able to run the above DLESyM example on CPU after installing earth2grid with the binary and then installing the DLESyM model.

https://github.com/NVlabs/earth2grid?tab=readme-ov-file

Binary release
Currently these are built for CUDA12, x86_64, linux, and pytorch 2.7 only.

VERSION=v2025.7.1
pip install \
  -f https://github.com/NVlabs/earth2grid/releases/expanded_assets/$VERSION \
  earth2grid

I had previously run into errors installing earth2grid via the method on the webpage...

pip install --no-build-isolation "earth2grid @ git+https://github.com/NVlabs/earth2grid@aefb10793aa372bae7d0951d627a89e2983fd0ca"

**Comment by pzharrington** [comment]
Nice you got it working! To clarify, you installed earth2grid on a CPU-only system with the binary, and then earth2studio with the `dlesym` extras, and the example ran? If so, that's interesting to know it works. As of now earth2grid does not seem to support builds (from source) on CPU-only systems, though once installed some of the ops can be run in CPU-only mode, which would explain why it runs for you. Maybe you circumvented the install limitation via the binary install. What were the set of install commands you used in this case?

**Comment by matthewjaniga** [comment]
Correct, install and running were done on CPU-only system.

conda create -n earth2studio python=3.12
conda activate earth2studio

pip install "earth2studio @ git+https://github.com/NVIDIA/earth2studio"

VERSION=v2025.7.1
pip install \
  -f https://github.com/NVlabs/earth2grid/releases/expanded_assets/$VERSION \
  earth2grid

pip install earth2studio[dlesym]

conda install cartopy

Then was able to run...

https://nvidia.github.io/earth2studio/_downloads/a271da38840794f3f45172882f237938/14_dlesym_example.py

**Comment by NickGeneva** [comment]
Thanks for follow up @matthewjaniga 

Good information for anyone wanting to use earth2grid on the CPU.

Will close this since seems things are resolved.
